### ISSUE TITLE
Inherited serializable fields lead to incorrect serialization order

### ISSUE THREAD

**Issue by IARI** [Kotlin/kotlinx.serialization#378]
This Issue is producable with kotlin >=1.3.20.

Consider the following scenario with a base class:

```kotlin
import kotlinx.serialization.Serializable

@Serializable
abstract class TestBase {
    val b = "val b"
    val a = "val a"
}
```
..and one (or two) implementations

```kotlin
@Serializable
class Test : TestBase() {
    val c = "val c"
}

@Serializable
class Test2 : TestBase() {
    val c2 = "val c2"
}
```

The serialized output of a "Test" object with `Json(indented = true)` is not consistent:
Sometimes it will be as expected, but sometimes it will be this:
```
{
    "a": "val b",
    "b": "val a",
    "c": "val c"
}
```

However, I was not able to get an absolutely clear order of file manipulation to yield the above output, but In order to reproduce the behavior, I would usually have to change something in the implementation class(es) `Test` or `Test2`, like remove or add the second implementation.
It also seemed to only work, when base- and implementation classes were in separate Files.

**Comment by sandwwraith** [comment]
The only workaround for now is, unfortunately, put them in one file.

**Comment by altavir** [comment]
Putting them in one project does not seem to help for now. Are there any ideas about how and when it could be fixed? I can use temporary fixes for now, but if there is no ETA, I will have to significantly change the structure of the project.

**Comment by IARI** [comment]
I keep getting wrong keys in the serialized json strings with inherited classes.

For instance I got

        {
            "nameProperty": "D:\\github\\42\\lambo-sbh",
            "versionProperty": "Unnamed",
            "nameProperty": "Lambo Sbh",
            "versionProperty": "0.0.1",
            ...
        }

Where the first two properties are inherited from a base class and should have other keys (not nameProperty, versionProperty).

inheritance to me seems broken in the serializer.
If there is any way in which i can assist with further investigating, please instruct me - but I believe, this should be considered a serious issue.

This feature is not production ready and either it should somehow be marked as such - or this issues should be understood and tackled.

**Comment by altavir** [comment]
The only workaround I found is to mark ancestor as non-serializeable and override all properties in the child. It does the trick, but requires a lot of duplicated code.

**Comment by IARI** [comment]
@altavir thanks for sharing this - However it defeats the point of inheritance a bit.
There are many cases where I used serializable abstract base classes with inherited properties that need to be serialized. 
Now I have to convert all the corresponding abstract base classes to interfaces.  To me it also seems that this is a bug that was only introduced with 1.3.20, (or maybe 1.3 already) because everything had worked fine before.

This still means that in this library Inheritance - which doesn't seem to be clearly documented anyways - is broken, and that should probably be mentioned somewhere in the readme.
Maybe in the part where it is talked about polymorphism ... that one is outdated anyways.

**Comment by IARI** [comment]
I have found another workaround.
Delegation does work for me, if the Object is instantiated by default in the constructor:

```
@Serializable
class MyImpl(val base: BaseClass = BaseClass()) : BaseClass by base {
    ...
}
```

**Comment by sandwwraith** [comment]
Unfortunately, this issue connected to the compiler implementation details (to be more specific, compiler data structures provide different order of properties in a class depending on the case where the class was analyzed – from source code or from compiled metadata), so I can't share ETA for now.

**Comment by IARI** [comment]
Thanks for sharing the information.

What about documenting, that inheritance of fields is basically broken in the serializer until then?

Also, is there a corresponding Issue in the compiler repo, that requests for a consistens order?
If so, coul'd you refer to ist - if not, could you create one (or alternatively can I create one)?

**Comment by altavir** [comment]
@sandwwraith I think it should be possible to make a workaround by manually defining the serialization order for fields. It would be helpful anyway to have this feature (I seldom want for JSON field to be in a specific order, and I need to rewrite class for that). It could be done as an optional part of descriptor and user interface could look like `KSerializer::withOrder(names: Set<String>): KSerializer`.

Currently, this problem makes a significant portion of the framework unusable.

**Comment by IARI** [comment]
Also keep in mind my first comment on this issue:
As it seems, in a derived class, not just
* the order of serialized field names is inconsist
but also additionally
* the name of inherited fields sometimes seems to be overwritten with names from fields of the deriving class.

I do not have any insight to judge if these issues are conected, but if they are not, the latter seems to be even more problematic to me. If it is helpful, i will create another issue for that, and get the problem down to a minimal reproducible example.

**Comment by IARI** [comment]
@sandwwraith 
first of all, thanks for all the good work on this lib in general.

Regarding this isse:
Judging from the references, it seems critical for several people,
So far there is not much information on the progress made I would be delighted, if you could give us an update.

- In case the underlying issue hasen't been tackled yet, would it be possible to ask the compiler-plugin team on a timeline for that?
- if that isn't feasible, at least, could a workaround as suggested by @altavir be implemented?

**Comment by sandwwraith** [comment]
This problem requires work on the compiler itself, not the compiler plugin. I've raised a discussion with compiler team about this issue some time ago, and we reached the conclusion that required changes can be done in 1.4-M2 timeframe, if everything goes well. So this would definitely be fixed in 1.4, and in some milestone releases before it.

**Comment by sandwwraith** [comment]
I can confirm that most of the bugs related to multi-module/project setups are fixed in Kotlin 1.4 (starting from 1.4-M3, to be precise). So this long and painful story is finally over.
